xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,244 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"errors"
+	"sync"
+
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+var errNotReceivedUpdate = errors.New("tried to construct a cluster update on a cluster that has not received an update")
+
+// clusterHandler will be given a name representing a cluster. It will then update the CDS policy
+// constantly with a list of Clusters to pass down to XdsClusterResolverLoadBalancingPolicyConfig
+// in a stream like fashion.
+type clusterHandler struct {
+	// A mutex to protect entire tree of clusters.
+	clusterMutex    sync.Mutex
+	root            *clusterNode
+	rootClusterName string
+
+	// A way to ping CDS Balancer about any updates or errors to a Node in the tree.
+	// This will either get called from this handler constructing an update or from a child with an error.
+	// Capacity of one as the only update CDS Balancer cares about is the most recent update.
+	updateChannel chan clusterHandlerUpdate
+
+	xdsClient xdsClientInterface
+}
+
+func (ch *clusterHandler) updateRootCluster(rootClusterName string) {
+	ch.clusterMutex.Lock()
+	defer ch.clusterMutex.Unlock()
+	if ch.root == nil {
+		// Construct a root node on first update.
+		ch.root = createClusterNode(rootClusterName, ch.xdsClient, ch)
+		ch.rootClusterName = rootClusterName
+		return
+	}
+	// Check if root cluster was changed. If it was, delete old one and start new one, if not
+	// do nothing.
+	if rootClusterName != ch.rootClusterName {
+		ch.root.delete()
+		ch.root = createClusterNode(rootClusterName, ch.xdsClient, ch)
+		ch.rootClusterName = rootClusterName
+	}
+}
+
+// This function tries to construct a cluster update to send to CDS.
+func (ch *clusterHandler) constructClusterUpdate() {
+	// If there was an error received no op, as this simply means one of the children hasn't received an update yet.
+	if clusterUpdate, err := ch.root.constructClusterUpdate(); err == nil {
+		// For a ClusterUpdate, the only update CDS cares about is the most recent one, so opportunistically drain the
+		// update channel before sending the new update.
+		select {
+		case <-ch.updateChannel:
+		default:
+		}
+		ch.updateChannel <- clusterHandlerUpdate{chu: clusterUpdate, err: nil}
+	}
+}
+
+// close() is meant to be called by CDS when the CDS balancer is closed, and it cancels the watches for every cluster in
+// the cluster tree.
+func (ch *clusterHandler) close() {
+	ch.clusterMutex.Lock()
+	defer ch.clusterMutex.Unlock()
+	ch.root.delete()
```

Comment:
OCD: set ch.root to `nil`, and ch.rootClusterName to `""`

Reply:
I will, but is there any engineering reason for this? Just wondering? If CDS closes does all this stuff get deallocated?

Reply:
You are right, this isn't strictly necessary. You can even argue that it's a waste of two value assignments.

It would matter if there's a way to _restart_ the clusterHander (to give it a new clusterName to watch, after Close())
This doesn't happen in CDS.

I like to keep the fields consistent. Like, if the handler isn't watching anything, it's root is nil, and it's cluster name is "". This is the case before the first watch, and should also be the state after the last watch (after Close()).

I'm OK either way. I will leave it to you.

Reply:
Sure, I'll do it your way. Thanks for the explanation.